opinions/united_states/state/ri_p.py

```diff
@@ -34,59 +34,126 @@ def __init__(self):
+                date_regex = r"^\(\w+\s+\d+\,\s+\d+\)$"
+                # Special case 1:
+                #   td1 contains only a date, matching date_regex
+                #   first line of summary contains case name with docket number
+                if re.match(date_regex, td1) and re.match(self.regex, text[0]):
+                    search = re.search(self.regex, text[0], re.MULTILINE)
+                    case_name = search.group(1)
+                    dockets = search.group(2)
+                    dockets = self._normalize_dockets(dockets)
+                    date_string = td1.strip()
+                    case_date = datetime.strptime(date_string,
+                                                  '(%B %d, %Y)').date()
+                    results.append({'date' : case_date,
+                                    'url' : href,
+                                    'docket' : dockets,
+                                    'summary' : "\n".join(text[1:]),
+                                    'case_name' : case_name})
+                else:
+                    # Special case 2:
+                    # td1 contains a very long case name
+                    #     docket number is missing
+                    #     date is missing
+                    link_text = tr1.xpath("./td[2]/a/text()")[0]
+                    # use link text as docket number
+                    link_text = self._normalize_dockets(link_text)
+                    # use date from previous record
+                    results.append({'date' : results[-1]['date'],
+                                    'url' : href,
+                                    'docket' : link_text,
+                                    'summary' : "\n".join(text),
+                                    'case_name' : case_name})
+            else:
+                search = re.search(self.regex, td1, re.MULTILINE)
+                case_name = search.group(1)
+                dockets = search.group(2)
+                dockets = self._normalize_dockets(dockets)
+                date_string = search.group(3)
+                if date_string:
+                    date_string = date_string.strip()
+                    # remove parentheses
+                    if date_string[0] == '(':
+                        date_string = date_string[1:]
+                    if date_string[-1] == ')':
+                        date_string = date_string[:-1]
+                    case_date = datetime.strptime(date_string, '%B %d, %Y').date()
+                else:
+                    # if no date, use date from previous record
+                    case_date = results[-1]['date']
+                results.append({'date' : case_date,
+                                'url' : href,
+                                'docket' : dockets,
+                                'summary' : "\n".join(text),
+                                'case_name' : case_name})
+        return results
+
+    def _filter(self, key):
```

Comment:
Two things here:

1. Unless I'm mistaken, the _load() function will be run every time that a `_get_*` function is run. Can we put in a cache so that we can hit that rather than reprocessing the data each time? 

1. Can we run something like this for `_filter`, to avoid the nested loop:

        for d in self._load():
            if d.get(key) is not None:
                r.append(d[key])